src/librustc_mir/borrow_check/universal_regions.rs

```diff
@@ -73,6 +74,37 @@ pub struct UniversalRegions<'tcx> {
     pub unnormalized_input_tys: &'tcx [Ty<'tcx>],
 
     pub yield_ty: Option<Ty<'tcx>>,
+
+    /// Extra information about region relationships, used
+    /// only when printing diagnostics.
+    ///
+    /// When processing closures/generators, we may generate multiple
+    /// region variables that all correspond to the same early-bound region.
+    /// We don't want to record these in `UniversalRegionRelations`,
+    /// as this would interfere with the propagation of closure
+    /// region constraints back to the parent function.
+    ///
+    /// Instead, we record this additional information here.
+    /// We map each region variable to a set of all other
+    /// region variables that correspond to the same early-bound region.
+    ///
+    /// For example, if we generate the following variables:
+    ///
+    /// 'a -> (_#0r, _#1r)
+    /// 'b -> (_#2r, _#3r)
+    ///
+    /// Then the map will look like this:
+    /// _#0r -> _#1r
+    /// _#1r -> _#0r
+    /// _#2r -> _#3r
+    /// _#3r -> _#2r
+    ///
+    /// When we compute upper bounds during diagnostic generation,
+    /// we accumulate a set of 'duplicate' from all non-duplicate
+    /// regions we've seen so far. Before we compute an upper bound,
+    /// we check if the region appears in our duplicates set - if so,
+    /// we skip it.
+    pub diagnostic_dup_regions: FxHashMap<RegionVid, FxHashSet<RegionVid>>,
```

Comment:
```suggestion
    pub diagnostic_dup_regions: RVarMapToEarlyBound,
```
(+ a type alias to use elsewhere)